signer-service/src/api/services/siwe.service.js

```diff
@@ -1,75 +1,139 @@
-    address: maybeSiweData.address,
-    message: maybeSiweData.siweMessage.toMessage(),
+    address: address,
+    message: siweMessage.toMessage(), // Validation must be done on the message as string
     signature,
   });
 
   if (!valid) {
-    throw new Error('Invalid signature.');
+    throw new ValidationError('Invalid signature');
+  }
+
+  // Perform additional checks to ensure message fields
+  if (siweMessage.nonce !== nonce) {
+    throw new ValidationError('Nonce mismatch');
+  }
+
+  if (siweMessage.expirationTime && new Date(siweMessage.expirationTime) < new Date()) {
+    throw new ValidationError('Message has expired');
+  }
+
+  return siweMessage;
+};
+
+// Since the message is created in the UI, we need to verify the fields of the message
+const verifyMessageFields = (siweMessage) => {
+  // Fields we validate on initial
+  const domain = siweMessage.domain;
+  const uri = siweMessage.uri;
+  const scheme = siweMessage.scheme; // must be https
+  const chainId = siweMessage.chainId;
+  const expirationTime = siweMessage.expirationTime;
+
+  if (!VALID_SIWE_LOGIN_ORIGINS.includes(uri)) {
+    throw new ValidationError('Origin not in the list of allowed origins');
   }
 
-  // Perform additional checks to ensure message integrity
-  if (maybeSiweData.siweMessage.nonce !== nonce) {
-    throw new Error('Nonce mismatch.');
+  if (!VALID_SIWE_DOMAINS.includes(domain)) {
+    throw new ValidationError('Incorrect domain');
   }
 
-  if (maybeSiweData.expirationTime && new Date(maybeSiweData.expirationTime) < new Date()) {
-    throw new Error('Message has expired.');
+  if (!VALID_SIWE_CHAINS.includes(chainId)) {
+    throw new ValidationError('Incorrect chain ID');
   }
 
-  // Successful verification, remove messages with same address
-  // from map except for the one we just verified.
+  if (scheme !== 'https') {
+    throw new ValidationError('Scheme must be https');
+  }
+
+  if (expirationTime === NaN || expirationTime === undefined) {
+    throw new ValidationError('Must defined valid expiration time');
+  }
+
+  // Check if expiration is within a reasonable range from current time
+  const currentTime = new Date().getTime();
+  const expirationTimestamp = new Date(expirationTime).getTime();
+  console.log('expirationTimestamp', expirationTimestamp);
+
+  const expirationGracePeriod = 1000 * 60; // 1 minute
+  const expirationPeriodMs = DEFAULT_LOGIN_EXPIRATION_TIME_HOURS * 60 * 60 * 1000;
+  const expectedMinExpirationTimestamp = currentTime + expirationPeriodMs - expirationGracePeriod;
+  const expectedMaxExpirationTimestamp = currentTime + expirationPeriodMs;
+
+  if (expirationTimestamp < expectedMinExpirationTimestamp) {
+    throw new ValidationError('Expiration time is too low');
+  }
+
+  if (expirationTimestamp > expectedMaxExpirationTimestamp) {
+    throw new ValidationError('Expiration time is too high');
+  }
+};
+
+const initialVerifySiweMessage = async (nonce, signature, siweMessage) => {
```

Comment:
Maybe rename to 
```suggestion
const verifyAndStoreSiweMessage = async (nonce, signature, siweMessage) => {
```
or similar? It's a bit difficult to see the logical difference between `verifySiweMessage` and `initialVerifySiweMessage` otherwise.

Reply:
Agreed. That will help the comments.